### PR TITLE
fix: [CDS-54849]: SSH key and WinRm secret reference is not updating while using in pipeline

### PR DESCRIPTION
**Pull request by ivan-83**
The issue is when the SSH key is added in the ShellScript step as the target host key in the pipeline, the SSH key entity reference is not updated to inform that the key is referenced by the pipeline.  

The code changes include adding WithSecretRef for referencing secrets in entities and adding referred secret entities in GenericStepPMSFilterJsonCreatorV2 

Screenshot 2023-04-07 at 00 00 24

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/46197)

### CODE REVIEW

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46197)* status: 4 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/filters/GenericStepPMSFilterJsonCreatorV2.java` line 107 at r2](https://reviewable.io/reviews/harness/harness-core/46197#-NSPzVw-EqNxzKYneoNC:-NSPzVw-EqNxzKYneoND:b-v0u67j) ([raw file](https://github.com/harness/harness-core/blob/505b1d933611caf97510e37ba322508034425101/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/filters/GenericStepPMSFilterJsonCreatorV2.java#L107)):*
> ```Java
>           + YAMLFieldNameConstants.SPEC + PATH_CONNECTOR + refKey;
>     } else if (entityType == EntityTypeProtoEnum.SECRETS) {
>       return fullQualifiedDomainNameFromNode + (isEmpty(fullQualifiedDomainNameFromNode) ? "" : PATH_CONNECTOR)
> ```

why spec is not path of FQN in secrets?

**Review by ivan-83 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46197)* status: 4 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/filters/GenericStepPMSFilterJsonCreatorV2.java` line 107 at r2](https://reviewable.io/reviews/harness/harness-core/46197#-NSPzVw-EqNxzKYneoNC:-NSQJi3sC20-thiO0IR8:bpuv0ng) ([raw file](https://github.com/harness/harness-core/blob/505b1d933611caf97510e37ba322508034425101/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/filters/GenericStepPMSFilterJsonCreatorV2.java#L107)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

why spec is not path of FQN in secrets?
</blockquote></details>

There is no 'spec' property in ShellScriptStepInfo. It is >> fullQualifiedDomainNameFromNode + "executionTarget.connectorRef" 
The 'executionTarget' property exists in ShellScriptStepInfo.

**Comment by ivan-83** [comment]
trigger messagecheck

**Review by ivan-83 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46197)* status: 12 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/filters/GenericStepPMSFilterJsonCreatorV2.java` line 107 at r2](https://reviewable.io/reviews/harness/harness-core/46197#-NSPzVw-EqNxzKYneoNC:-NSc261o0USU5EK4C1Du:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/505b1d933611caf97510e37ba322508034425101/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/filters/GenericStepPMSFilterJsonCreatorV2.java#L107)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

There is no 'spec' property in ShellScriptStepInfo. It is >> fullQualifiedDomainNameFromNode + "executionTarget.connectorRef" 
The 'executionTarget' property exists in ShellScriptStepInfo.
</blockquote></details>

Done.

File: pipeline-service/modules/orchestration-steps/contracts/src/main/java/io/harness/steps/shellscript/ShellScriptStepInfo.java [hunk 1]
```diff
@@ -85,4 +88,14 @@ public SpecParameters getSpecParameters() {
   public ParameterField<List<TaskSelectorYaml>> fetchDelegateSelectors() {
     return getDelegateSelectors();
   }
+
+  @Override
+  public Map<String, ParameterField<String>> extractSecretRefs() {
```
Comment (acasian-harness) [inline pipeline-service/modules/orchestration-steps/contracts/src/main/java/io/harness/steps/shellscript/ShellScriptStepInfo.java:93]: aren't we missing the same for command step when it is configured to run ondelegate?

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/infra/yaml/SshWinRmAzureInfrastructure.java [hunk 2]
```diff
@@ -162,4 +163,9 @@ public Map<String, ParameterField<String>> extractConnectorRefs() {
     connectorRefMap.put(YAMLFieldNameConstants.CONNECTOR_REF, connectorRef);
```
Comment (acasian-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/infra/yaml/SshWinRmAzureInfrastructure.java:163]: we are checking for each infra if `connectorRef != null`, why are we missing this for azure?

**Review by ivan-83 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46197)* status: 13 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/infra/yaml/SshWinRmAzureInfrastructure.java` line 163 at r4](https://reviewable.io/reviews/harness/harness-core/46197#gh-1161710937:-NSfJWCf9ta5B2uPgJxF:bqp5966) ([raw file](https://github.com/harness/harness-core/blob/7c4e70212cd6d6093b177d248e4b87b212745354/127-cd-nextgen-entities/src/main/java/io/harness/cdng/infra/yaml/SshWinRmAzureInfrastructure.java#L163)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

we are checking for each infra if `connectorRef != null`, why are we missing this for azure?
</blockquote></details>

There is @NotNull annotation on connectorRef for Azure and AWS and it's not needed. However, there is flow "Specify hosts" or PDC  where connectorRef can be null.
___
*[`pipeline-service/modules/orchestration-steps/contracts/src/main/java/io/harness/steps/shellscript/ShellScriptStepInfo.java` line 93 at r4](https://reviewable.io/reviews/harness/harness-core/46197#gh-1161710835:-NSfK2oW4Ha8A-_wGdDE:bs3844t) ([raw file](https://github.com/harness/harness-core/blob/7c4e70212cd6d6093b177d248e4b87b212745354/pipeline-service/modules/orchestration-steps/contracts/src/main/java/io/harness/steps/shellscript/ShellScriptStepInfo.java#L93)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

aren't we missing the same for command step when it is configured to run ondelegate?
</blockquote></details>

Yes, will update the code.

**Review by ivan-83 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46197)* status: 13 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration-steps/contracts/src/main/java/io/harness/steps/shellscript/ShellScriptStepInfo.java` line 93 at r4](https://reviewable.io/reviews/harness/harness-core/46197#gh-1161710835:-NSfRcbN9g7VXxhEh0Z8:bma41bl) ([raw file](https://github.com/harness/harness-core/blob/7c4e70212cd6d6093b177d248e4b87b212745354/pipeline-service/modules/orchestration-steps/contracts/src/main/java/io/harness/steps/shellscript/ShellScriptStepInfo.java#L93)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

Yes, will update the code. 
</blockquote></details>

@acasian-harness 
I've investigated "Run on Delegate" and we are supplying SSH/WinRM credentials on infra definition, not as a part of the pipeline. I've already implemented logic to link credentials used in infra to infra.

**Review by acasian-harness — APPROVED** [review]

**Review by archit-harness — APPROVED** [review]
> Reviewed 9 of 9 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46197)* status: 13 file(s) to review. 2. unresolved discussions.

**Comment by ivan-83** [comment]
trigger review

**Pull request merged**